File: suc/kotlin/067.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CenterAlignedTopAppBar
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Shapes
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.Typography
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat

private const val NAME = "112*ClientDetail*en"
private const val UI_TYPE = "Dashboard"
private const val STYLE_THEME = "Warm Gradient"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFFFF7043)
        val secondary = Color(0xFFFFB74D)
        val tertiary = Color(0xFFFFD54F)
        val background = Color(0xFFFFF8E1)
        val surface = Color(0xFFFFFFFF)
        val surfaceVariant = Color(0xFFFFECB3)
        val outline = Color(0xFFE0C097)
        val success = Color(0xFF43A047)
        val warning = Color(0xFFFBC02D)
        val error = Color(0xFFE53935)
        val onPrimary = Color(0xFFFFFFFF)
        val onSecondary = Color(0xFF3E2723)
        val onTertiary = Color(0xFF3E2723)
        val onBackground = Color(0xFF3E2723)
        val onSurface = Color(0xFF3E2723)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 28.sp, fontWeight = FontWeight.Bold)
        val headline = TextStyle(fontSize = 20.sp, fontWeight = FontWeight.SemiBold)
        val title = TextStyle(fontSize = 16.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
    object Shapes {
        val small = RoundedCornerShape(8.dp)
        val medium = RoundedCornerShape(12.dp)
        val large = RoundedCornerShape(16.dp)
    }
    object Spacing {
        val xs = 4.dp
        val sm = 8.dp
        val md = 12.dp
        val lg = 16.dp
        val xl = 24.dp
        val xxl = 32.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(1.dp, 4.dp, 2.dp, 0.1f)
        val level2 = ShadowSpec(3.dp, 8.dp, 4.dp, 0.14f)
        val level3 = ShadowSpec(6.dp, 12.dp, 6.dp, 0.16f)
    }
}

private val AppColorScheme = lightColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline,
    error = AppTokens.Colors.error
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    headlineMedium = AppTokens.TypographyTokens.headline,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

data class Client(val id: Int, val name: String, val region: String, val level: String, val x: Float, val y: Float)

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RootScreen() {
    val clients = listOf(
        Client(1, "Sunrise Tech", "Shanghai", "A", 0.2f, 0.3f),
        Client(2, "Delta Logistics", "Beijing", "B", 0.5f, 0.45f),
        Client(3, "Oceanic Foods", "Guangzhou", "A", 0.72f, 0.6f),
        Client(4, "Horizon Ltd.", "Chengdu", "C", 0.33f, 0.75f)
    )
    val selectedId = remember { mutableStateOf<Int?>(null) }
    Scaffold(
        contentWindowInsets = WindowInsets(0),
        topBar = {
            CenterAlignedTopAppBar(
                title = {
                    Text(text = "Client Detail", style = MaterialTheme.typography.displayLarge, color = MaterialTheme.colorScheme.onSurface)
                }
            )
        },
        containerColor = MaterialTheme.colorScheme.background
    ) { padding ->
        Column(
            modifier = Modifier.fillMaxSize().padding(padding).padding(horizontal = AppTokens.Spacing.lg, vertical = AppTokens.Spacing.md),
            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md)
        ) {
            Box(
                modifier = Modifier.fillMaxWidth().height(200.dp).background(
                    Brush.linearGradient(listOf(AppTokens.Colors.secondary, AppTokens.Colors.primary)),
                    AppTokens.Shapes.large
                ),
                contentAlignment = Alignment.Center
            ) {
                Text(text = "Regional Client Map", style = MaterialTheme.typography.headlineMedium, color = AppTokens.Colors.onPrimary)
            }
            BoxWithConstraints(
                modifier = Modifier.fillMaxWidth().heightIn(min = 360.dp).aspectRatio(0.9f).background(AppTokens.Colors.surface, AppTokens.Shapes.large).border(1.dp, AppTokens.Colors.outline, AppTokens.Shapes.large)
            ) {
                MapGridOverlay()
                clients.forEach { c ->
                    MapMarker(
                        label = c.name,
                        level = c.level,
                        selected = selectedId.value == c.id,
                        onClick = { selectedId.value = if (selectedId.value == c.id) null else c.id },
                        modifier = Modifier.offset(
                            x = maxWidth * c.x,
                            y = maxHeight * c.y
                        )
                    )
                }
            }
            LazyColumn(verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md), contentPadding = PaddingValues(bottom = AppTokens.Spacing.xxl)) {
                items(clients) { c ->
                    Card(
                        modifier = Modifier.fillMaxWidth(),
                        shape = AppTokens.Shapes.medium,
                        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surface),
                        elevation = CardDefaults.cardElevation(defaultElevation = AppTokens.ElevationMapping.level1.elevation),
                        border = BorderStroke(1.dp, MaterialTheme.colorScheme.outline)
                    ) {
                        Row(
                            modifier = Modifier.fillMaxWidth().padding(AppTokens.Spacing.lg),
                            horizontalArrangement = Arrangement.SpaceBetween,
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Column(verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.xs)) {
                                Text(text = c.name, style = MaterialTheme.typography.titleMedium, color = MaterialTheme.colorScheme.onSurface)
                                Text(text = c.region, style = MaterialTheme.typography.bodyMedium, color = MaterialTheme.colorScheme.onSurface)
                            }
                            Text(text = "Level ${c.level}", style = MaterialTheme.typography.labelMedium, color = MaterialTheme.colorScheme.primary)
                        }
                    }
                }
            }
        }
    }
}

@Composable
fun BoxScope.MapMarker(label: String, level: String, selected: Boolean, onClick: () -> Unit, modifier: Modifier) {
    val color = when (level) {
        "A" -> AppTokens.Colors.primary
        "B" -> AppTokens.Colors.secondary
        else -> AppTokens.Colors.tertiary
    }
    Box(
        modifier = modifier.size(if (selected) 28.dp else 22.dp).background(color, CircleShape).border(2.dp, if (selected) AppTokens.Colors.onPrimary else AppTokens.Colors.surface, CircleShape).align(Alignment.TopStart)
    )
    Card(
        onClick = onClick,
        modifier = modifier.offset(x = 20.dp, y = (-4).dp),
        shape = AppTokens.Shapes.small,
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceVariant),
        elevation = CardDefaults.cardElevation(defaultElevation = AppTokens.ElevationMapping.level1.elevation),
        border = BorderStroke(1.dp, MaterialTheme.colorScheme.outline)
    ) {
        Text(text = label, style = MaterialTheme.typography.labelMedium, color = MaterialTheme.colorScheme.onSurface, modifier = Modifier.padding(horizontal = AppTokens.Spacing.sm, vertical = AppTokens.Spacing.xs))
    }
}

@Composable
fun MapGridOverlay() {
    Column(modifier = Modifier.fillMaxSize().padding(AppTokens.Spacing.lg), verticalArrangement = Arrangement.SpaceEvenly) {
        repeat(6) {
            Box(modifier = Modifier.fillMaxWidth().height(1.dp).background(AppTokens.Colors.surfaceVariant))
        }
    }
    Row(modifier = Modifier.fillMaxSize().padding(AppTokens.Spacing.lg), horizontalArrangement = Arrangement.SpaceEvenly) {
        repeat(6) {
            Box(modifier = Modifier.width(1.dp).fillMaxSize().background(AppTokens.Colors.surfaceVariant))
        }
    }
}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.navigationBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.navigationBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFFFFFFF)
@Composable
fun PreviewRoot() {
    AppTheme {
        RootScreen()
    }
}

File: suc/swift/067.swift
import SwiftUI

// MARK: - AppTokens
// This struct defines the design tokens for the application, making it easy to modify styles globally.
struct AppTokens {
    // Colors used throughout the application.
    struct Colors {
        static let primary = Color(red: 0xFF / 255.0, green: 0x70 / 255.0, blue: 0x43 / 255.0)
        static let secondary = Color(red: 0xFF / 255.0, green: 0xB7 / 255.0, blue: 0x4D / 255.0)
        static let tertiary = Color(red: 0xFF / 255.0, green: 0xD5 / 255.0, blue: 0x4F / 255.0)
        static let background = Color(red: 0xFF / 255.0, green: 0xF8 / 255.0, blue: 0xE1 / 255.0)
        static let surface = Color(red: 0xFF / 255.0, green: 0xFF / 255.0, blue: 0xFF / 255.0)
        static let surfaceVariant = Color(red: 0xFF / 255.0, green: 0xEC / 255.0, blue: 0xB3 / 255.0)
        static let outline = Color(red: 0xE0 / 255.0, green: 0xC0 / 255.0, blue: 0x97 / 255.0)
        static let success = Color(red: 0x43 / 255.0, green: 0xA0 / 255.0, blue: 0x47 / 255.0)
        static let warning = Color(red: 0xFB / 255.0, green: 0xC0 / 255.0, blue: 0x2D / 255.0)
        static let error = Color(red: 0xE5 / 255.0, green: 0x39 / 255.0, blue: 0x35 / 255.0)
        static let onPrimary = Color(red: 0xFF / 255.0, green: 0xFF / 255.0, blue: 0xFF / 255.0)
        static let onSecondary = Color(red: 0x3E / 255.0, green: 0x27 / 255.0, blue: 0x23 / 255.0)
        static let onTertiary = Color(red: 0x3E / 255.0, green: 0x27 / 255.0, blue: 0x23 / 255.0)
        static let onBackground = Color(red: 0x3E / 255.0, green: 0x27 / 255.0, blue: 0x23 / 255.0)
        static let onSurface = Color(red: 0x3E / 255.0, green: 0x27 / 255.0, blue: 0x23 / 255.0)
    }

    // Typography styles for different text elements.
    struct TypographyTokens {
        static let display = Font.system(size: 28, weight: .bold)
        static let headline = Font.system(size: 20, weight: .semibold)
        static let title = Font.system(size: 16, weight: .medium)
        static let body = Font.system(size: 14, weight: .regular)
        static let label = Font.system(size: 12, weight: .medium)
    }

    // Shapes with predefined corner radii.
    struct Shapes {
        static let small = RoundedRectangle(cornerRadius: 8)
        static let medium = RoundedRectangle(cornerRadius: 12)
        static let large = RoundedRectangle(cornerRadius: 16)
    }

    // Standardized spacing values.
    struct Spacing {
        static let xs: CGFloat = 4
        static let sm: CGFloat = 8
        static let md: CGFloat = 12
        static let lg: CGFloat = 16
        static let xl: CGFloat = 24
        static let xxl: CGFloat = 32
    }

    // Shadow specifications for elevation effects.
    struct ShadowSpec {
        let elevation: CGFloat // Corresponds to blur radius in SwiftUI's shadow
        let radius: CGFloat    // Not directly mapped, often combined with elevation
        let dy: CGFloat        // Y-offset for the shadow
        let opacity: Double    // Opacity of the shadow
    }

    // Predefined elevation levels with their shadow specs.
    struct ElevationMapping {
        static let level1 = ShadowSpec(elevation: 1, radius: 4, dy: 2, opacity: 0.1)
        static let level2 = ShadowSpec(elevation: 3, radius: 8, dy: 4, opacity: 0.14)
        static let level3 = ShadowSpec(elevation: 6, radius: 12, dy: 6, opacity: 0.16)
    }
}

// MARK: - Custom Environment Values for Theming
// These structs and extensions allow us to mimic a MaterialTheme-like system in SwiftUI
// by providing custom color schemes and typography through the environment.

// Defines the application's color scheme.
struct AppColorScheme {
    let primary: Color
    let onPrimary: Color
    let secondary: Color
    let onSecondary: Color
    let background: Color
    let onBackground: Color
    let surface: Color
    let onSurface: Color
    let surfaceVariant: Color
    let outline: Color
    let error: Color
}

// Key for accessing the AppColorScheme in the environment.
private struct AppColorSchemeKey: EnvironmentKey {
    static let defaultValue = AppColorScheme(
        primary: AppTokens.Colors.primary,
        onPrimary: AppTokens.Colors.onPrimary,
        secondary: AppTokens.Colors.secondary,
        onSecondary: AppTokens.Colors.onSecondary,
        background: AppTokens.Colors.background,
        onBackground: AppTokens.Colors.onBackground,
        surface: AppTokens.Colors.surface,
        onSurface: AppTokens.Colors.onSurface,
        surfaceVariant: AppTokens.Colors.surfaceVariant,
        outline: AppTokens.Colors.outline,
        error: AppTokens.Colors.error
    )
}

extension EnvironmentValues {
    var appColorScheme: AppColorScheme {
        get { self[AppColorSchemeKey.self] }
        set { self[AppColorSchemeKey.self] = newValue }
    }
}

// Defines the application's typography.
struct AppTypography {
    let displayLarge: Font
    let headlineMedium: Font
    let titleMedium: Font
    let bodyMedium: Font
    let labelMedium: Font
}

// Key for accessing the AppTypography in the environment.
private struct AppTypographyKey: EnvironmentKey {
    static let defaultValue = AppTypography(
        displayLarge: AppTokens.TypographyTokens.display,
        headlineMedium: AppTokens.TypographyTokens.headline,
        titleMedium: AppTokens.TypographyTokens.title,
        bodyMedium: AppTokens.TypographyTokens.body,
        labelMedium: AppTokens.TypographyTokens.label
    )
}

extension EnvironmentValues {
    var appTypography: AppTypography {
        get { self[AppTypographyKey.self] }
        set { self[AppTypographyKey.self] = newValue }
    }
}

// A ViewModifier to apply the custom theme to any view hierarchy.
struct AppThemeModifier: ViewModifier {
    func body(content: Content) -> some View {
        content
            .environment(\.appColorScheme, AppColorSchemeKey.defaultValue)
            .environment(\.appTypography, AppTypographyKey.defaultValue)
            // Shapes are typically applied directly via modifiers like .clipShape or .overlay
            // The global background is set in the RootScreen for full-screen coverage.
    }
}

extension View {
    // Convenience method to apply the custom theme.
    func appTheme() -> some View {
        self.modifier(AppThemeModifier())
    }
}

// MARK: - Client Data Model
// Represents a client with various details and map coordinates.
struct Client: Identifiable {
    let id: Int
    let name: String
    let region: String
    let level: String
    let x: CGFloat // Relative X position (0.0 to 1.0)
    let y: CGFloat // Relative Y position (0.0 to 1.0)
}

// MARK: - Custom Views

// A custom view for displaying a map marker with a label.
struct MapMarker: View {
    let label: String
    let level: String
    let selected: Bool
    let onClick: () -> Void
    let xOffset: CGFloat // Base X offset from the container's top-leading corner
    let yOffset: CGFloat // Base Y offset from the container's top-leading corner

    @Environment(\.appColorScheme) private var colorScheme
    @Environment(\.appTypography) private var typography

    var body: some View {
        // ZStack is used to layer the marker dot and the label card.
        // Alignment is .topLeading, and then individual offsets are applied.
        ZStack(alignment: .topLeading) {
            let markerColor = {
                switch level {
                case "A": return AppTokens.Colors.primary
                case "B": return AppTokens.Colors.secondary
                default: return AppTokens.Colors.tertiary
                }
            }()
            
            // Marker Dot (CircleShape in Compose)
            Circle()
                .fill(markerColor)
                .frame(width: selected ? 28 : 22, height: selected ? 28 : 22)
                .overlay(
                    Circle()
                        .stroke(selected ? AppTokens.Colors.onPrimary : AppTokens.Colors.surface, lineWidth: 2)
                )
                .offset(x: xOffset, y: yOffset) // Apply the base offset to the marker dot

            // Label Card (Card in Compose)
            Button(action: onClick) {
                Text(label)
                    .font(typography.labelMedium)
                    .foregroundColor(colorScheme.onSurface)
                    .padding(.horizontal, AppTokens.Spacing.sm)
                    .padding(.vertical, AppTokens.Spacing.xs)
                    .background(colorScheme.surfaceVariant)
                    .clipShape(AppTokens.Shapes.small)
                    .overlay(
                        AppTokens.Shapes.small
                            .stroke(colorScheme.outline, lineWidth: 1)
                    )
                    .shadow(color: Color.black.opacity(AppTokens.ElevationMapping.level1.opacity),
                            radius: AppTokens.ElevationMapping.level1.radius, // blur radius
                            x: 0,
                            y: AppTokens.ElevationMapping.level1.dy) // y-offset
            }
            .buttonStyle(PlainButtonStyle()) // Remove default button styling for custom appearance
            // Apply the base offset plus the additional offset specified in Kotlin
            .offset(x: xOffset + 20, y: yOffset - 4) 
        }
    }
}

// A custom view to draw a grid overlay on the map.
// Mimics `Arrangement.SpaceEvenly` for 6 lines within the padded area.
struct MapGridOverlayCorrected: View {
    @Environment(\.appColorScheme) private var colorScheme

    var body: some View {
        GeometryReader { geometry in
            // Calculate available space for the grid lines after applying padding.
            let availableWidth = geometry.size.width - 2 * AppTokens.Spacing.lg
            let availableHeight = geometry.size.height - 2 * AppTokens.Spacing.lg

            ZStack {
                // Horizontal lines: 6 lines distributed evenly, meaning 7 equal segments.
                ForEach(0..<6) { i in
                    Rectangle()
                        .fill(colorScheme.surfaceVariant)
                        .frame(height: 1)
                        // Offset from the top-leading corner of the GeometryReader.
                        // Add AppTokens.Spacing.lg for the top padding.
                        // (i + 1) because the first line is at the end of the first segment.
                        .offset(y: AppTokens.Spacing.lg + (availableHeight / 7.0) * CGFloat(i + 1))
                }

                // Vertical lines: 6 lines distributed evenly, meaning 7 equal segments.
                ForEach(0..<6) { i in
                    Rectangle()
                        .fill(colorScheme.surfaceVariant)
                        .frame(width: 1)
                        // Offset from the top-leading corner of the GeometryReader.
                        // Add AppTokens.Spacing.lg for the left padding.
                        // (i + 1) because the first line is at the end of the first segment.
                        .offset(x: AppTokens.Spacing.lg + (availableWidth / 7.0) * CGFloat(i + 1))
                }
            }
        }
    }
}

// MARK: - RootScreen
// The main content view of the application.
struct RootScreen: View {
    @State private var selectedId: Int? = nil // State to track the selected client marker.
    @Environment(\.appColorScheme) private var colorScheme
    @Environment(\.appTypography) private var typography

    // Sample client data.
    let clients = [
        Client(id: 1, name: "Sunrise Tech", region: "Shanghai", level: "A", x: 0.2, y: 0.3),
        Client(id: 2, name: "Delta Logistics", region: "Beijing", level: "B", x: 0.5, y: 0.45),
        Client(id: 3, name: "Oceanic Foods", region: "Guangzhou", level: "A", x: 0.72, y: 0.6),
        Client(id: 4, name: "Horizon Ltd.", region: "Chengdu", level: "C", x: 0.33, y: 0.75)
    ]

    var body: some View {
        // VStack acts as the Scaffold, arranging Top Bar and content vertically.
        VStack(spacing: 0) {
            // Custom Top Bar (mimics CenterAlignedTopAppBar)
            ZStack {
                Text("Client Detail")
                    .font(typography.displayLarge)
                    .foregroundColor(colorScheme.onSurface)
                    .padding(.vertical, AppTokens.Spacing.md) // Vertical padding for text
            }
            .frame(maxWidth: .infinity)
            .frame(height: 56) // Standard Material3 TopAppBar height
            .background(colorScheme.background) // Background color for the top bar area
            
            // Scrollable content area (mimics LazyColumn and other content below TopAppBar)
            ScrollView {
                VStack(spacing: AppTokens.Spacing.md) { // Vertical arrangement with spacing
                    // Box 1: Regional Client Map header
                    Text("Regional Client Map")
                        .font(typography.headlineMedium)
                        .foregroundColor(AppTokens.Colors.onPrimary)
                        .frame(maxWidth: .infinity, minHeight: 200) // height(200.dp)
                        .background(
                            LinearGradient(
                                gradient: Gradient(colors: [AppTokens.Colors.secondary, AppTokens.Colors.primary]),
                                startPoint: .leading, // Corresponds to Brush.linearGradient default behavior
                                endPoint: .trailing
                            )
                        )
                        .clipShape(AppTokens.Shapes.large) // RoundedCornerShape(large)
                        .padding(.horizontal, AppTokens.Spacing.lg) // Horizontal padding from Column
                        .padding(.top, AppTokens.Spacing.md) // Top padding from Column

                    // BoxWithConstraints: Map area
                    GeometryReader { geometry in
                        ZStack(alignment: .topLeading) { // Alignment for MapMarker positioning
                            AppTokens.Shapes.large // Background shape
                                .fill(colorScheme.surface)
                                .overlay(
                                    AppTokens.Shapes.large // Border
                                        .stroke(colorScheme.outline, lineWidth: 1)
                                )
                            
                            MapGridOverlayCorrected() // The grid overlay

                            // Iterate through clients to place MapMarkers
                            ForEach(clients) { client in
                                MapMarker(
                                    label: client.name,
                                    level: client.level,
                                    selected: selectedId == client.id,
                                    onClick: {
                                        selectedId = (selectedId == client.id) ? nil : client.id
                                    },
                                    // Calculate marker position based on relative coordinates and GeometryReader size
                                    xOffset: geometry.size.width * client.x,
                                    yOffset: geometry.size.height * client.y
                                )
                            }
                        }
                        // Apply frame and aspect ratio to the ZStack within GeometryReader
                        .frame(maxWidth: .infinity)
                        .aspectRatio(0.9, contentMode: .fit) // aspectRatio(0.9f)
                        .frame(minHeight: 360) // heightIn(min = 360.dp)
                    }
                    // Apply frame and aspect ratio to the GeometryReader itself to define its bounds
                    .frame(maxWidth: .infinity)
                    .aspectRatio(0.9, contentMode: .fit)
                    .frame(minHeight: 360)
                    .padding(.horizontal, AppTokens.Spacing.lg) // Horizontal padding from Column

                    // LazyColumn: Client List
                    VStack(spacing: AppTokens.Spacing.md) { // Vertical arrangement with spacing
                        ForEach(clients) { client in
                            // Card for each client
                            ZStack {
                                AppTokens.Shapes.medium // Card background shape
                                    .fill(colorScheme.surface)
                                    .shadow(color: Color.black.opacity(AppTokens.ElevationMapping.level1.opacity),
                                            radius: AppTokens.ElevationMapping.level1.radius,
                                            x: 0,
                                            y: AppTokens.ElevationMapping.level1.dy)
                                    .overlay(
                                        AppTokens.Shapes.medium // Card border
                                            .stroke(colorScheme.outline, lineWidth: 1)
                                    )
                                
                                HStack { // Row for client details
                                    VStack(alignment: .leading, spacing: AppTokens.Spacing.xs) { // Column for name and region
                                        Text(client.name)
                                            .font(typography.titleMedium)
                                            .foregroundColor(colorScheme.onSurface)
                                        Text(client.region)
                                            .font(typography.bodyMedium)
                                            .foregroundColor(colorScheme.onSurface)
                                    }
                                    Spacer() // Pushes "Level" text to the right
                                    Text("Level \(client.level)")
                                        .font(typography.labelMedium)
                                        .foregroundColor(colorScheme.primary)
                                }
                                .padding(AppTokens.Spacing.lg) // Padding inside the card
                            }
                            .frame(maxWidth: .infinity) // Card fills available width
                        }
                    }
                    .padding(.horizontal, AppTokens.Spacing.lg) // Horizontal padding from Column
                    .padding(.bottom, AppTokens.Spacing.xxl) // contentPadding for LazyColumn
                }
            }
            .background(colorScheme.background) // Background for the scrollable content
        }
        .ignoresSafeArea(.all) // Make the app content extend to all edges, ignoring safe areas.
        .statusBarHidden(true) // Hide the top status bar.
        .background(colorScheme.background) // Ensure the very root background matches the theme.
    }
}

// MARK: - App Entry Point
// The main entry point for the SwiftUI application.
@main
struct ClientDetailApp: App {
    var body: some Scene {
        WindowGroup {
            RootScreen()
                .appTheme() // Apply the custom theme to the entire app.
        }
    }
}